Code/Tools/TestImpactFramework/Runtime/Native/Code/Source/TestRunner/Native/Job/TestImpactNativeShardedTestJobInfoGenerator.h

```diff
@@ -0,0 +1,329 @@
+    }
+
+    template<typename TestJobRunner>
+    auto NativeShardedTestRunJobInfoGeneratorBase<TestJobRunner>::GenerateJobInfos(
+        const AZStd::vector<TestTargetAndEnumeration>& testTargetsAndEnumerations)
+        -> AZStd::vector<ShardedTestJobInfo<TestJobRunner>>
+    {
+        AZStd::vector<ShardedTestJobInfo<TestJobRunner>> jobInfos;
+        for (size_t testTargetIndex = 0, jobId = 0; testTargetIndex < testTargetsAndEnumerations.size(); testTargetIndex++)
+        {
+            jobInfos.push_back(GenerateJobInfo(testTargetsAndEnumerations[testTargetIndex], { jobId }));
+            jobId += jobInfos.back().GetJobInfos().size();
+        }
+
+        return jobInfos;
+    }
+
+    template<typename TestJobRunner>
+    RepoPath NativeShardedTestRunJobInfoGeneratorBase<TestJobRunner>::GenerateShardedTargetRunArtifactFilePath(
+        const NativeTestTarget* testTarget, const size_t shardNumber) const
+    {
+        auto artifactFilePath = GenerateTargetRunArtifactFilePath(testTarget, m_artifactDir.m_shardedTestRunArtifactDirectory);
+        return artifactFilePath.ReplaceExtension(AZStd::string::format("%zu%s", shardNumber, artifactFilePath.Extension().String().c_str()).c_str());
+    }
+
+    template<typename TestJobRunner>
+    AZStd::string NativeShardedTestRunJobInfoGeneratorBase<TestJobRunner>::GenerateShardedLaunchCommand(
+        const NativeTestTarget* testTarget, const RepoPath& shardAdditionalArgsFile) const
+    {
+        return AZStd::string::format(
+            "%s --args_from_file \"%s\"",
+            GenerateLaunchArgument(testTarget, m_targetBinaryDir, m_testRunnerBinary).c_str(),
+            shardAdditionalArgsFile.c_str());
+    }
+
+    template<typename TestJobRunner>
+    RepoPath NativeShardedTestRunJobInfoGeneratorBase<TestJobRunner>::GenerateShardedAdditionalArgsFilePath(
+        const NativeTestTarget* testTarget, size_t shardNumber) const
+    {
+        return AZStd::string::format(
+            "%s.%zu.args", (m_artifactDir.m_shardedTestRunArtifactDirectory / RepoPath(testTarget->GetName())).c_str(), shardNumber);
+    }
+} // namespace TestImpact
+
+namespace AZStd
+{
+    //! Less function for TestTargetAndEnumeration types for use in maps and sets.
+    template<>
+    struct less<TestImpact::TestTargetAndEnumeration>
+    {
+        bool operator()(const TestImpact::TestTargetAndEnumeration& lhs, const TestImpact::TestTargetAndEnumeration& rhs) const
+        {
+            return reinterpret_cast<size_t>(lhs.first) < reinterpret_cast<size_t>(rhs.first);
+        }
+    };
+
+    //! Hash function for TestTargetAndEnumeration types for use in unordered maps and sets.
+    template<>
+    struct hash<TestImpact::TestTargetAndEnumeration>
+    {
+        size_t operator()(const TestImpact::TestTargetAndEnumeration& testTargetAndEnumeration) const noexcept
+        {
+            return reinterpret_cast<size_t>(testTargetAndEnumeration.first);
+        }
+    };
```

Comment:
This specializations cost a lot more compile time than adding a member function to the class that is `operator<` and a `operator size_t`.

Plus the TestTargetAndEnumeration `operator<` can be implemented in a cpp file, saving compile time
```suggestion
    bool TestImpact::TestTargetAndEnumeration::operator<(const TestImpact::TestTargetAndEnumeration& other) const
    {
        return reinterpret_cast<size_t>(this->first) < reinterpret_cast<size_t>(other.first);
    }

    //! Hash function for TestTargetAndEnumeration types for use in unordered maps and sets.
    size_t TestImpact::operator size_t() const noexcept
    {
        return reinterpret_cast<size_t>(this->first);
    }
```

Reply:
I'll leave this comment as unresolved and will address in the staging branch as this will span more than this PR 👍